### PR TITLE
Spinning food model

### PR DESCRIPTION
**Pull request by otsha**
## New functionality
- A 3d model of a bowl of ramen is rendered on the front page
  - By default, the bowl alternates between slightly rotating clockwise and slightly rotating counter-clockwise
  - While the lottery is in progress, the bowl spins around counter-clockwise
  - When the lottery results are shown, the bowl is hidden until the lottery is restarted

### Notes
- The p5 script can be found in `packages/app/src/scripts/food.js`
- The code is well commented to explain what is happening - for more information, refer to [the p5.js reference](https://p5js.org/reference)
- When the restaurant result is shown, the canvas's CSS `display` property is set to `none` to avoid re-loading the assets for every consecutive lottery shuffle
- When leaving the front page, the canvas is destroyed - this means that the assets will have to be loaded again when returning to the front page
- The assets are served from `packages/app/src/data/assets/`
- Basic tests check that the canvas is visible/hidden when it should be.

## Administrative tasks
- Created a small wrapper to avoid repeated code when creating p5 canvases
- Moved all p5 related React components to `components/p5Components/`
- Created a `scripts/` directory to the front-end root for p5 scripts - after all, they are not services and should not be in the `services/` directory
- Updated the `#root` element's background color to match the animation's background. **This was done purely to make the visual outlook unified for now, the colours will likely be adjusted later on.**

### CODE REVIEW

**Review by skoskipaa — COMMENTED** [review]
The spinning bowl looks tasty and code is ok too!

File: packages/app/src/scripts/food.js [hunk 1]
```diff
@@ -0,0 +1,214 @@
+import bowl from '../data/assets/bowl.obj'
+import bowlRim from '../data/assets/bowlrim.obj'
+import soup from '../data/assets/soup.obj'
+import salad from '../data/assets/salad.obj'
+import noodles from '../data/assets/noodles-lowpoly.obj'
+import chopsticks from '../data/assets/chopsticks.obj'
+import eggWhite from '../data/assets/eggwhite.obj'
+import eggYolk from '../data/assets/eggyellow.obj'
+import tofu from '../data/assets/tofu.obj'
```
Comment (ritinae) [inline packages/app/src/scripts/food.js:9]: I personally don't think this is a good idea. Use textured .obj instead?

File: packages/app/src/scripts/food.js [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    /* Set up appropriate positioning and scaling */
+
+    sketch.scale(100)
+    sketch.translate(0, 1, 0)
+    sketch.rotateX(170)
+    sketch.rotateZ(0)
+
+    spinning
+      ? sketch.rotateY(rotation += Math.abs(Math.sin(spinningTime / 18)) * 10)
+      : sketch.rotateY(rotation += Math.sin(sketch.frameCount / 8))
```
Comment (ritinae) [inline packages/app/src/scripts/food.js:126]: "Configuration" variables for colors and animations are sprinkled all over the file, mostly as magic numbers. While very common practice in the context of strictly graphics related code *(which tends to be "hacky")*, it is not very good for maintainability on the long run. I propose all configuration related variables and magic numbers *(where sensible)* are collected to named properties or constants and moved to the top of the file.

File: packages/app/src/scripts/food.js [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+      /* Draw the different parts of the model*/
+      drawModel('bowl', materials.white)
+      drawModel('bowlRim', materials.blue)
+      drawModel('soup', materials.soup)
+      drawModel('salad', materials.salad)
+      drawModel('noodles', materials.noodle)
+      drawModel('chopsticks', materials.wood)
+      drawModel('eggWhite', materials.white)
+      drawModel('eggYolk', materials.eggYolk)
+      drawModel('tofu', materials.white)
```
Comment (ritinae) [inline packages/app/src/scripts/food.js:142]: Having some background in writing performant rendering code, this gives me the shivers. We are making a draw call for each part of a model. WebGL draw calls *(or typically draw/submit calls in any rendering framework)* are ridiculously expensive and here we are making nine of them per frame for a single model. We should seriously consider packing the model in some more practical form. *(e.g. single mesh with UV-mapped texture)*

File: packages/app/src/scripts/food.js [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    sketch.background(bgColor)
+
+    /* Ambient lighting based on the background color helps tie the models to the scene */
+
+    let ambientStrength = 0.9
+    sketch.ambientLight(
+      sketch.red(bgColor) * ambientStrength,
+      sketch.green(bgColor) * ambientStrength,
+      sketch.blue(bgColor) * ambientStrength
+    )
```
Comment (ritinae) [inline packages/app/src/scripts/food.js:109]: This doesn't do anything? The material functions override the ambient light anyway?

**Review by ritinae — DISMISSED** [review]

**Comment by ritinae** [comment]
Well, everything looks OK. I just don't like the way a single model has to be split to 9 parts for coloring things with different colors. See [proof-of-concept](https://github.com/team-lunch-app/lunch-app/blob/poc-bowl-textured/packages/app/src/scripts/food.js) of flattened, textured model on how this could be handled in the model instead of in the code.

What I've done there is:
 1. I imported the model to blender and joined all parts into a single object with multiple materials.
 2. Changed to blender's image editor view. There I created a new black texture with size 1024x1024 pixels.
 2. Changed to the shader editor and assigned a dummy image texture node (not connected to anything but selected) with the image created in step 2. selected. This setup is the very unintuitive way of telling blender where to paint the colors of those materials later on. Note that the selection of the image texture node remains even after you change to another material or view.
 3. Now, I marked seams on the model and UV unwrapped it
 4. When happy with UV layout, I changed the renderer from EEVEE to Cycles and baked the diffuse colors. For this, I selected diffuse from the dropdown and selected only colors to be baked. (Make sure to select all image texture nodes in individual materials' shader editors and the model to be baked. Then go to render tab, disable selected to active and hit bake) As we "assigned" our target image to materials in step 2., the colors have now magically appeared on our image texture.
 5. Created clone of the model and cleared its materials, created a single new material and assigned it the generated image texture.
 6. Presto! Now we have a textured model with a single material.
 7. P5.js for some reason uses inverted UV-coordinates so the image still needs to be vertically flipped in image editor before saving to file.
 8. Now, load the single model and the texture with P5.js and assign the texture when rendering.

Notes:
 - The UV layout is really lazy on the POC-branch, the noodles take the most of the texture space even though they could very well be overlaid on top of each other to make manual edits with Photoshop more pleasant experience.
 - Having to manually flip the image texture is kind of annoying but I didn't find a quick-fix for this from P5 docs. There most likely is some configuration flag or property somewhere, but I didn't find it.
 - We lose the ability to adjust specularity on per-part basis. Either everything is specular or nothing is specular. For refined specularity adjustment, a separate specularity map should be baked. (Not sure if p5 supports this. Likely not.)
 - The texture is an image. An image can be painted to with any image manipulation software. This allows more complex graphics on the bowl, for example, if desired.

File: packages/app/src/scripts/food.js (hunk 4 above)
Comment (ritinae) [reply packages/app/src/scripts/food.js:109]: This indeed *does* do things. Curious.

**Review by ritinae — APPROVED** [review]

**Pull request merged**